File: src/main/java/at/dietze/quadru/events/PlayerSitOnStairs.java
package at.dietze.quadru.events;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.type.Stairs;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;

public class PlayerSitOnStairs implements Listener {

    @EventHandler
    public void playerSitOnStairs(PlayerInteractEvent e) {
        if (e.getClickedBlock() == null) return;
        if (e.getAction() != Action.RIGHT_CLICK_BLOCK) return;
        Player p = e.getPlayer();

        if (p.isSneaking()) return;
        if (p.getInventory().getItemInMainHand().getType() != Material.AIR) return;

        Block block = e.getClickedBlock();

        Block aboveBlock = block.getRelative(BlockFace.UP);
        if (!aboveBlock.isEmpty()) return;

        BlockData blockData = block.getBlockData();

        if (!(blockData instanceof Stairs)) return;

        Stairs stairs = (Stairs) blockData;
        if (stairs.getShape() != Stairs.Shape.STRAIGHT) return;

        BlockFace facing = stairs.getFacing();

        Location spawnLocation = block.getLocation().add(0.5, 0.25, 0.5);

        switch (facing) {
            case NORTH:
                spawnLocation.add(0, 0, 0.25);
                break;
            case SOUTH:
                spawnLocation.add(0, 0, -0.25);
                break;
            case EAST:
                spawnLocation.add(-0.25, 0, 0);
                break;
            case WEST:
                spawnLocation.add(0.25, 0, 0);
                break;
        }

        Entity seat = p.getWorld().spawnEntity(spawnLocation, EntityType.ARROW);

        seat.setGravity(false);
        seat.setInvulnerable(true);
        seat.setSilent(true);
        seat.setPersistent(false);

        seat.addPassenger(p);
    }
}

File: src/main/java/at/dietze/quadru/events/PlayerSitOnStairs.kt
package at.dietze.quadru.events

import org.bukkit.Material
import org.bukkit.block.BlockFace
import org.bukkit.block.data.type.Stairs
import org.bukkit.entity.EntityType
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent

class PlayerSitOnStairs : Listener {
    @EventHandler
    fun playerSitOnStairs(e: PlayerInteractEvent) {
        if (e.clickedBlock == null) return
        if (e.action != Action.RIGHT_CLICK_BLOCK) return
        val p = e.player

        if (p.isSneaking) return
        if (p.inventory.itemInMainHand.type != Material.AIR) return

        val block = e.clickedBlock

        val aboveBlock = block!!.getRelative(BlockFace.UP)
        if (!aboveBlock.isEmpty) return

        val blockData = block.blockData as? Stairs ?: return

        val stairs = blockData
        if (stairs.shape != Stairs.Shape.STRAIGHT) return

        val facing = stairs.facing

        val spawnLocation = block.location.add(0.5, 0.25, 0.5)

        when (facing) {
            BlockFace.NORTH -> spawnLocation.add(0.0, 0.0, 0.25)
            BlockFace.SOUTH -> spawnLocation.add(0.0, 0.0, -0.25)
            BlockFace.EAST -> spawnLocation.add(-0.25, 0.0, 0.0)
            BlockFace.WEST -> spawnLocation.add(0.25, 0.0, 0.0)
            BlockFace.UP -> TODO()
            BlockFace.DOWN -> TODO()
            BlockFace.NORTH_EAST -> TODO()
            BlockFace.NORTH_WEST -> TODO()
            BlockFace.SOUTH_EAST -> TODO()
            BlockFace.SOUTH_WEST -> TODO()
            BlockFace.WEST_NORTH_WEST -> TODO()
            BlockFace.NORTH_NORTH_WEST -> TODO()
            BlockFace.NORTH_NORTH_EAST -> TODO()
            BlockFace.EAST_NORTH_EAST -> TODO()
            BlockFace.EAST_SOUTH_EAST -> TODO()
            BlockFace.SOUTH_SOUTH_EAST -> TODO()
            BlockFace.SOUTH_SOUTH_WEST -> TODO()
            BlockFace.WEST_SOUTH_WEST -> TODO()
            BlockFace.SELF -> TODO()
        }

        val seat = p.world.spawnEntity(spawnLocation, EntityType.ARROW)

        seat.setGravity(false)
        seat.isInvulnerable = true
        seat.isSilent = true
        seat.isPersistent = false

        seat.addPassenger(p)
    }
}
